content/blog/chess-knight-problem/index.md

```diff
@@ -0,0 +1,129 @@
+
+So, right now, we have all of the squares labelled that we can get to in zero, one or two moves. If you would like, you can fill in three moves and possibly four moves if that is required to fill every square. Remember to give precedent to the lower number in any already occupied squares.
+
+If any of the labelled squares is a desired destination, then we know the minimum number of moves required to reach that square. So, all we have to do is start with our starting square and repeat this process until we happen to fill our ending destination with a number. The number in that square will be the minimal number of moves required to reach that spot.
+
+# JavaScript Execution {#js-execution}
+
+So, let's get started. I hacked this together in codepen, and I didn't build an interface for it, but that would be an easy enough step. We could do all kinds of animations in D3js, etc, but that's not for this blog post.
+
+To begin, let's define a two dimensional array to be our chess board:
+
+```javascript
+const board = [];
+for (let i = 0; i < 8; i++) {
+  board[i] = [];
+}
+```
+
+That's it. All we need is a two dimensional 8x8 array. I've not explicitly defined the number of entries in the either dimension of the array because that's one of the idiosyncrasies of JavaScript. (It would be possible to say something like _let board = Array(8);_ but it's not clear how much that will benefit performance here - JavaScript is famous for not having the same memory management optimisation of languages like FORTRAN - also note that I didn't need to prepopulate the array with null values, as I would have to with many other languages).
+
+Let's make a function to add a move to the board. We need to see if the location is both in range and not occupied by another number before adding that move:
+
+```javascript
+const addMove = (x, y, level) => {
+  if ((x >= 0) && (x <= 7) && (y >= 0) && (y <= 7) && board[x][y] == null) {
+    board[x][y] = level;
+  }
+}
+```
+
+You may have noticed that I have checked if _board\[x\]\[y\] == null_, rather than just _!board\[x\]\[y\]_. This is because _0_ is a potential entry for one of the squares (our staring square) and _0_ is falsy.
+
+Let's call this from another function that for any given location, tries to add all of the moves for that can be reached from that location:
+
+```javascript
+const addAllMoves = (x, y, level) => {
+  addMove(x + 1, y + 2, level);
+  addMove(x + 2, y + 1, level);
+  addMove(x + 2, y - 1, level);
+  addMove(x + 1, y - 2, level);
+  addMove(x - 1, y - 2, level);
+  addMove(x - 2, y - 1, level);
+  addMove(x - 2, y + 1, level);
+  addMove(x - 1, y + 2, level);
+}
+```
+
+There might be a way to do this systematically, but for what I was doing it was easier to tap it out by hand. Notice that we didn't have to check for valid moves, because that is handled by _addMove()_.
```

Comment:
```suggestion
There might be a way to do this systematically, but for what I was doing, it was easier to tap it out by hand. Notice that we didn't have to check for valid moves because that is handled by `addMove()`.
```